### ISSUE TITLE
Error on ABF read with AxonIO

### ISSUE THREAD

**Issue by dreamerinmadness** [NeuralEnsemble/python-neo#253]
Hi!

When i tried to read an ABF file that i made on Clampex 8.2 and later modified with Clampfit 10.6.2, I get the next error:

`File "test.py", line 12, in <module>
    bl = r.read_block()
  File "/usr/lib/python2.7/dist-packages/neo/io/axonio.py", line 157, in read_block
    header = self.read_header()
  File "/usr/lib/python2.7/dist-packages/neo/io/axonio.py", line 424, in read_header
    ADCInfo['ADCChNames'] = strings[ADCInfo['lADCChannelNameIndex']-1]
IndexError: list index out of range`

Checking over the code in axonio.py, I found that the error was 'cause "goodstart" in read_header() look for "compex" or "axoscope", but my file has none of both

`SSCH##X##AXENGN 2.0.2.2ClampfitC:\Axon\Params\neuromodulacion\Ach\Apl NMDA 2 s -80mV cada 2ms esp 10 2480us.proIN 1pACmd 0mVCmd 1mVC:\Documents and Settings\Equipo2\Escritorio\ruidos\puroruido1.abfAO #2mVAO #3mVAO #4mVAO #5mVAO #6mVAO #7mV`

So, I added:
`if goodstart == -1 :
                goodstart = bigString.lower().find('clampfit') - 1`

Just after of:
`goodstart = bigString.lower().find('clampex')
            if goodstart == -1 :
                goodstart = bigString.lower().find('axoscope')`

With that change my problem was solved, but I don't know if that's the better approach to really solve it.

The same error (solved with the same add to axonio.py code) is yielded when i try to read any of my files saved on clampfit (mean currents of multiple sweeps). If you want it, can found one of this files here:
[https://mega.nz/#!QhUiyQxZ!ecDuSoZ2mWbcJYKTUIU7QxdpcLZJhGr22-AIwYO6-R4](https://mega.nz/#!QhUiyQxZ!ecDuSoZ2mWbcJYKTUIU7QxdpcLZJhGr22-AIwYO6-R4)

(Sorry if something is misspelled, i'ḿ from Mexico and my English is out of practice)

**Comment by samuelgarcia** [comment]
hi,
thank you for this.

I have made a patch PR #254 .
Could you try it ?

I also put one of your files (renamed File_axon_7.abf) on gnode repo for testing.

**Comment by dreamerinmadness** [comment]
Well, works fine, except that 'ADCChNames' and 'ADCChUnits' don't coincide with header, so returns a dimensionless array:

`[[-4.20613098]
 [-2.43611145]
 [ 0.9818573 ]
 ..., 
 [-7.99031067]
 [-6.64753723]
 [-9.57722473]] dimensionless`

I mean, 'strings' returns:

`['Clampfit', 'C:\\Axon\\Params\\neuromodulacion\\Ach\\Apl NMDA 3 s -80mV cada 2ms esp 15 2480us.pro', 'Current', 'pA', 'Cmd 0', 'mV', 'Cmd 1', 'mV', 'AO #2', 'mV', 'AO #3', 'mV', '']`

Where first element ('Clampfit') is the software where the file was made, the second ('C:\\Axon\\Params\\neuromodulacion\\Ach\\Apl NMDA 3 s -80mV cada 2ms esp 15 2480us.pro') the protocol file used, and later gives the first chanel name ('Current'), their units (pA), the second chanel name ('Cmd 0') and units ('mV'), and so on. So, if 'lADCChannelNameIndex' is 4 (-1, because index starts on 0) and 'lADCUnitsIndex' (-1), 'ADCChNames' and  'ADCChUnits' gonna be 'pA' and 'Cmd0' (respectively) when suppous to be 'Current' and 'pA'. That's the reason that I added a '-1' at the end of the "correction" that a send on my previous comment ( 'if goodstart == -1 : goodstart = bigString.lower().find('clampfit') - 1' ); if i added in same way a '-1' in the new code, the name and units corresponde, so 'quantities' recognice the units and adds it to the array, but I don't know if it happens when 'key' is another than 'Clampfit'.

So, I try your patch and thats what i find, hope it is usefull.

P.S: By the way, I was using the neo version that is on Debian Testing, recently started to use the versión on github (to be really usefull sending my coments like this), and i realized that each value on segments is now an independent array in place to be a single array with all values, I mean, now each segment is a multidimensional array, thats the way is gonna be from now? I'm asking to start to edit my scripts if this is the way is gonna be now

**Comment by samuelgarcia** [comment]
Hi.
Thank you for testing.
This is very strange.

The problem come from that in this file ```ADCInfo['lADCChannelNameIndex']``` and ```ADCInfo['lADCUnitsIndex']```. They are the index in ```strings``` list but one based.

So the problem is here:
```
                ADCInfo['ADCChNames'] = strings[ADCInfo['lADCChannelNameIndex'] - 1]
                ADCInfo['ADCChUnits'] = strings[ADCInfo['lADCUnitsIndex'] - 1]
```

In all other files they are odd and even and that match the ```strings``` list
In your file theses values are even and odd and that do not match.

The -1 you mention for goodstart is where the big_string starts so it is not related to that.
I guess it solve the problem because it enlarge the list with one empty element but it looks like uggly.

I have written that code long time ago and don't remember where I get the spec of the ABF file.

Do you have more file for testing with diffrents channel maps and channel order ?

**Comment by samuelgarcia** [comment]
My guess is that the new key is 'AXENGN' and not 'Clampfit'.
I have put the patch that should be OK.
Could you try it with some files that have complicated channel maps and check units and channel names.


For the second question, Andrew have made a big refactoring in June 2016 on neo object model.
Previsously (<0.5) AnalogSignal was dim=1 for only one signal and AnalogSignalArray was dim=2 for some signal with same sample rate. In case of same sample rate for all channels, each IO could implement either a list of N AnalogSignal shape(length,) (case for AxonIO) or 1 AnalogSignalArray shape (length, N).

Now (>=0.5, the master branch), AnalogSignal and AnalogSignalArray have been merged, so for historical reason AxonIO now propose a list of AnalogSignal with shape (length, 1).
It could make sense to rewrite this IO to get only one AnalogSignal with shape (length, N).

So you certainly migth change some part of your analysis, code if you use the new neo (>=0.5).
You can also still work with neo 0.4.x and propose a PR in that branch to get old neo schema.

**Comment by dreamerinmadness** [comment]
Thank you very much, whith the new patch everithing woks pretty well, as you say looks like the new kay is in deed 'AXENGN'.

About test with more files with different channel maps and channel order; I tested it with files recorded in two different channels, but is the best i can do, because in my lab we use old equipment that just support pClamp V8 and 9, so all my files are originally abf1, is when I analyze my data when use pClamp 10.6 and get some files whith abf2 as that i submitted you. Beside, by the moment, and for a while, I'm just using the kind of recordings and analysis that you seen on my file, but if anybody have some different files I can test any thing you want.

Finally, I just wanna reiterate my thanks for all your help, and offer my help for any thing you need (I mean it).

**Comment by samuelgarcia** [comment]
OK. Thank you for testing.
I merge the #254 .

**Comment by zm711** [comment]
Fixed by #254.
